File: material-stepper/src/test/java/com/stepstone/stepper/internal/widget/pagetransformer/StepperRtlPageTransformerTest.java
package com.stepstone.stepper.internal.widget.pagetransformer;

import android.view.View;

import com.stepstone.stepper.test.runner.StepperRobolectricTestRunner;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;

import static org.mockito.AdditionalMatchers.or;
import static org.mockito.Matchers.eq;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

/**
 * @author Piotr Zawadzki
 */
@RunWith(StepperRobolectricTestRunner.class)
public class StepperRtlPageTransformerTest {

    private static final int VIEW_WIDTH = 320;

    @Mock
    View mockView;

    StepperRtlPageTransformer pageTransformer = new StepperRtlPageTransformer();

    @Before
    public void setUp() {
        when(mockView.getWidth()).thenReturn(VIEW_WIDTH);
    }

    @Test
    public void should_reset_translationX_when_at_front_center() {
        //when
        pageTransformer.transformPage(mockView, 0.0f);

        //then
        verify(mockView).setTranslationX(isZero());
    }

    @Test
    public void should_reverse_view_when_at_one_full_position_to_the_right() {
        //when
        pageTransformer.transformPage(mockView, 1.0f);

        //then
        verify(mockView).setTranslationX(-2 * VIEW_WIDTH);
    }

    @Test
    public void should_reverse_view_when_at_one_full_position_to_the_left() {
        //when
        pageTransformer.transformPage(mockView, -1.0f);

        //then
        verify(mockView).setTranslationX(2 * VIEW_WIDTH);
    }

    @Test
    public void should_reverse_view_when_in_the_middle() {
        //when
        pageTransformer.transformPage(mockView, 0.5f);

        //then
        verify(mockView).setTranslationX(-VIEW_WIDTH);
    }

    private float isZero() {
        return or(eq(0.0f), eq(-0.0f));
    }

}

File: material-stepper/src/test/java/com/stepstone/stepper/internal/widget/pagetransformer/StepperRtlPageTransformerTest.kt
package com.stepstone.stepper.internal.widget.pagetransformer

import android.view.View
import com.nhaarman.mockito_kotlin.doReturn
import com.nhaarman.mockito_kotlin.mock
import org.junit.Test
import org.mockito.AdditionalMatchers.or
import org.mockito.ArgumentMatchers.eq
import org.mockito.Mockito.verify

/**
 * @author Piotr Zawadzki
 */
class StepperRtlPageTransformerTest {

    companion object {

        const val VIEW_WIDTH = 320

    }

    val mockView: View = mock {
        on { width } doReturn VIEW_WIDTH
    }

    val pageTransformer = StepperRtlPageTransformer()

    @Test
    fun `Should reset 'translationX' when page is at front center`() {
        //when
        pageTransformer.transformPage(mockView, 0.0f)

        //then
        verify(mockView).translationX = isZero
    }

    @Test
    fun `Should reverse view position on X axis when page is at one full position to the right`() {
        //when
        pageTransformer.transformPage(mockView, 1.0f)

        //then
        verify(mockView).translationX = -2.0f * VIEW_WIDTH
    }

    @Test
    fun `Should reverse view position on X axis when page is at one full position to the left`() {
        //when
        pageTransformer.transformPage(mockView, -1.0f)

        //then
        verify(mockView).translationX = 2.0f * VIEW_WIDTH
    }

    @Test
    fun `Should reverse view position on X axis when page is half moved`() {
        //when
        pageTransformer.transformPage(mockView, 0.5f)

        //then
        verify(mockView).translationX = -VIEW_WIDTH.toFloat()
    }

    val isZero: Float
        get() = or(eq(0.0f), eq(-0.0f))

}